tests/test_h1_server.c

```diff
@@ -0,0 +1,409 @@
+static void s_test_on_complete(struct aws_http_stream *stream, int error_code, void *user_data) {
+    struct tester_request *request = user_data;
+    (void)stream;
+    struct aws_http_stream *r_handler = request->request_handler;
+    struct aws_byte_buf storage_buf = request->storage;
+
+    request->method = r_handler->incoming_request_method_str;
+    request->method.ptr = storage_buf.buffer + storage_buf.len;
+    aws_byte_buf_write_from_whole_cursor(&storage_buf, r_handler->incoming_request_method_str);
+
+    request->uri = r_handler->incoming_request_uri;
+    request->uri.ptr = storage_buf.buffer + storage_buf.len;
+    aws_byte_buf_write_from_whole_cursor(&storage_buf, r_handler->incoming_request_uri);
+
+    // aws_http_stream_release(r_handler);
+    if (error_code == AWS_ERROR_SUCCESS) {
+    }
+}
+
+void s_tester_on_request_header(
+    struct aws_http_stream *stream,
+    const struct aws_http_header *header_array,
+    size_t num_headers,
+    void *user_data) {
+    (void)stream;
+    struct tester_request *request = user_data;
+    struct aws_byte_buf *storage = &request->storage;
+    const struct aws_http_header *in_header = header_array;
+    struct aws_http_header *my_header = request->headers + request->num_headers;
+    for (size_t i = 0; i < num_headers; ++i) {
+        /* copy-by-value, then update cursors to point into permanent storage */
+        *my_header = *in_header;
+
+        my_header->name.ptr = storage->buffer + storage->len;
+        AWS_FATAL_ASSERT(aws_byte_buf_write_from_whole_cursor(storage, in_header->name));
+
+        my_header->value.ptr = storage->buffer + storage->len;
+        AWS_FATAL_ASSERT(aws_byte_buf_write_from_whole_cursor(storage, in_header->value));
+
+        in_header++;
+        my_header++;
+        request->num_headers++;
+    }
+    // int index = request->num_headers;
+    // request->headers[index];
+}
+
+void s_tester_on_request_header_block_done(struct aws_http_stream *stream, bool has_body, void *user_data) {
+    (void)stream;
+    struct tester_request *request = user_data;
+    AWS_FATAL_ASSERT(request->header_done == false);
+    request->header_done = true;
+    request->has_incoming_body = has_body;
+}
+
+void s_tester_on_request_body(
+    struct aws_http_stream *stream,
+    const struct aws_byte_cursor *data,
+    size_t *out_window_update_size,
+    void *user_data) {
+    (void)out_window_update_size;
+    (void)stream;
+    struct tester_request *request = user_data;
+
+    AWS_FATAL_ASSERT(request->header_done == true);
+
+    AWS_FATAL_ASSERT(request->has_incoming_body);
+
+    /* Copy data into storage, and point body cursor at that */
+    if (!request->body.ptr) {
+        request->body.ptr = request->storage.buffer + request->storage.len;
+    }
+    request->body.len += data->len;
+
+    if (request->stop_auto_window_update) {
+        *out_window_update_size = 0;
+    }
+    AWS_FATAL_ASSERT(aws_byte_buf_write_from_whole_cursor(&request->storage, *data));
```

Comment:
Please move this write above the `request->body` modifications for reasons mentioned previously